src/librustc/ty/relate.rs

```diff
@@ -226,12 +226,16 @@ impl<'tcx> Relate<'tcx> for ty::ProjectionTy<'tcx> {
         where R: TypeRelation<'a, 'gcx, 'tcx>, 'gcx: 'a+'tcx, 'tcx: 'a
     {
         let tcx = relation.tcx();
-        if a.item_name(tcx) != b.item_name(tcx) {
+        if a.item_def_id != b.item_def_id {
+            // FIXME(tschottdorf): is this (still) the correct error?
             Err(TypeError::ProjectionNameMismatched(
                 expected_found(relation, &a.item_name(tcx), &b.item_name(tcx))))
         } else {
-            let trait_ref = relation.relate(&a.trait_ref, &b.trait_ref)?;
-            Ok(ty::ProjectionTy::from_ref_and_name(tcx, trait_ref, a.item_name(tcx)))
+            let substs = relation.relate(&a.substs, &b.substs)?;
```

Comment:
The `None` vs `Some` parameter is interesting. *However*, `relate_item_substs` for traits returns `None` and it might make sense to never *even pretend* to compute variance for traits. cc @nikomatsakis (but what you did is most likely fine either way)